### DISCUSSION TITLE
Custom Filter Operation

### DISCUSSION THREAD

**Discussion by MarkitoAgnito** [ChilliCream/hotchocolate#3779] [Q&A]
Hi, I want to add an additional custom filter operation which i would use to manipulate the filter text before execution of the query.

So where we have out of the box something like:
```
{
    myEntity(where: { myProperty_starts_with:"someText" })
    {
        nodes
        {
             myProperty
        }
    }
}
```

I want to be able to expose the following:
```
{
    myEntity(where: { myProperty_nysiis:"someText" })
    {
        nodes
        {
             myProperty
        }
    }
}
```

Im my service I have: 

```
services.AddGraphQLServer()                  
.AddQueryType(d => d.Name("Query"))
.AddConvention<IFilterConvention, CustomFilterConvention>()
.AddFiltering< CustomFilterConvention>()
```

my custom operation id:

```
    public static class CustomOperations
    {
        public const int Nysiis = 1025;
    } 
```

Filter convention plus handler:

```
    public class CustomFilterConvention : FilterConvention
    {
        public override NameString GetTypeName(Type runtimeType) =>
            base.GetTypeName(runtimeType) + "_nysiis";

        protected override void Configure(IFilterConventionDescriptor descriptor)
        {
            descriptor.AddDefaultOperations();
            descriptor.AddDefaults();

            descriptor 
                 .Operation(CustomOperations.Nysiis)
                .Name("Nysiis"); 

            descriptor
                .Configure<StringOperationFilterInputType>(_ => _
                    .Operation(CustomOperations.Nysiis)
                    .Type<StringType>());

            descriptor.AddProviderExtension(
                    new QueryableFilterProviderExtension(
                        y =>
                        {
                            y.AddDefaultFieldHandlers();
                            y.AddFieldHandler<QueryableStringNysiisHandler>();
                        }));
        }
    }

    public class QueryableStringNysiisHandler : QueryableStringOperationHandler
    {
        private static readonly MethodInfo _toLower = typeof(string)
            .GetMethods()
            .Single(
                _ => _.Name == nameof(string.ToLower) &&
                _.GetParameters().Length == 0); 

        protected override int Operation => CustomOperations.Nysiis; 

        public override Expression HandleOperation(
            QueryableFilterContext context,
            IFilterOperationField field,
            IValueNode value,
            object parsedValue)
        {
            var property = context.GetInstance();
            var str = (string)parsedValue;

            //operate on passed in string
            //str = do something with str
            return Expression.Equal(
                Expression.Call(property, _toLower),
                Expression.Constant(str.ToLower()));
        }
    }
```

Im not sure im doing this right as all i see now  instead of my properties is 'scope_<existing filter options> - now i dont see my properties nor the additional '_nysiss' one im trying to add.

Any direction welcome!

**Comment by PascalSenn** [comment]
it seems like you are using HotChocolate.Types.Filtering 
custom operations are only supported woth the new filtering in HotChocolate.Data

**Reply by MarkitoAgnito** [reply]
Thanks @PascalSenn this working fine now - I didnt realise the query syntax would change slightly where previously i would use:

{
    myEntity(where: { myProperty_contains:"someText" })
    {
        nodes
        {
             myProperty
        }
    }
}

and now it's:

{
    myEntity(where: { myProperty { nysiis:"someText" }})
    {
        nodes
        {
             myProperty
        }
    }
}

I'll look into seeing if i can keep existing syntax but at least i have it working, thanks!

**Reply by PascalSenn** [reply]
it would be 'where: {myproperty:{nyiis:"foo"}}'

**Reply by viktor-nikolaev** [reply]
Hi @PascalSenn 
what should I add if I want to customize ListFilterInput<FilterInputType<SomeDomainModel>>?

I try to use this, but no luck

       x.Configure<ListFilterInputType<FilterInputType<PackageView>>>(_ => _
                  .Operation(CustomOperations.Nysiis)
                  .Type<StringType>());
